extensions/amp-story/1.0/amp-story-tooltip.js

```diff
@@ -0,0 +1,387 @@
+      this.onTooltipStateUpdate_(el);
+    });
+  }
+
+  /**
+   * Builds the tooltip overlay and appends it to the provided story.
+   * @private
+   */
+  build_() {
+    this.isBuilt_ = true;
+
+    this.shadowRoot_ = this.win_.document.createElement('div');
+
+    this.tooltipOverlayEl_ =
+      dev().assert(this.buildTemplate_(this.win_.document));
+    createShadowRootWithStyle(this.shadowRoot_, this.tooltipOverlayEl_, CSS);
+
+    this.tooltipOverlayEl_
+        .addEventListener('click', event => this.onOutsideClickableEls_(event));
+
+    this.storeService_.subscribe(StateProperty.UI_STATE, isDesktop => {
+      this.onUIStateUpdate_(isDesktop);
+    }, true /** callToInitialize */);
+
+    this.storeService_.subscribe(StateProperty.CURRENT_PAGE_ID, () => {
+      // Hide active tooltip when page switch is triggered by keyboard or
+      // desktop buttons.
+      if (this.storeService_.get(StateProperty.TOOLTIP_ELEMENT)) {
+        this.closeTooltip_();
+      }
+    });
+
+    return this.shadowRoot_;
+  }
+
+  /**
+   * Hides the tooltip layer.
+   * @private
+   */
+  closeTooltip_() {
+    this.storeService_.dispatch(Action.TOGGLE_TOOLTIP, null);
+  }
+
+  /**
+   * Reacts to store updates related to tooltip active status.
+   * @param {!Element} clickedEl
+   * @private
+   */
+  onTooltipStateUpdate_(clickedEl) {
+    if (!clickedEl) {
+      this.resources_.mutateElement(dev().assertElement(this.tooltipOverlayEl_),
+          () => {
+            this.tooltipOverlayEl_
+                .classList.toggle('i-amphtml-hidden', true);
+          });
+      return;
+    }
+
+    if (!this.isBuilt_) {
+      this.storyEl_.appendChild(this.build_());
+    }
+
+    if (clickedEl != this.clickedEl_) {
+      this.clickedEl_ = clickedEl;
+      this.attachTooltipToEl_(clickedEl);
+    }
+
+    this.resources_.mutateElement(dev().assertElement(this.tooltipOverlayEl_),
+        () => {
+          this.tooltipOverlayEl_
+              .classList.toggle('i-amphtml-hidden', false);
+        });
+  }
+
+  /**
+   * Reacts to desktop state updates and hides navigation buttons since we
+   * already have in the desktop UI.
+   * @param {!UIType} uiState
+   * @private
+   */
+  onUIStateUpdate_(uiState) {
+    this.resources_.mutateElement(dev().assertElement(this.tooltipOverlayEl_),
+        () => {
+          uiState === UIType.DESKTOP ?
+            this.tooltipOverlayEl_.setAttribute('desktop', '') :
+            this.tooltipOverlayEl_.removeAttribute('desktop');
+        });
+  }
+
+  /**
+   * Builds tooltip and attaches it depending on the clicked element content and
+   * position.
+   * @param {!Element} clickedEl
+   * @private
+   */
+  attachTooltipToEl_(clickedEl) {
+    const {href} = parseUrlDeprecated(clickedEl.getAttribute('href'));
+
+    const domainName = getSourceOriginForElement(clickedEl, href);
+    const tooltipText =
+      clickedEl.getAttribute('data-tooltip-text') || domainName;
```

Comment:
We need to keep in mind that this won't work if the element is not a link, or if there's no `data-tooltip-text`. We should later consider adding a translated string like "Tap to open" as a fallback.

Since `getSourceOriginForElement` is doing quite a lot of stuff (getting services etc), could you try to write this differently to make sure it's only called if needed?

`const tooltipText = clickedEl.getAttribute('data-tooltip-text') || getSourceOriginForElement(clickedEl, href);`